packages/create-component-with-subscriptions/src/createComponentWithSubscriptions.js

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import React from 'react';
+import invariant from 'fbjs/lib/invariant';
+
+// TODO The below Flow types don't work for `Props`
+export function createComponent<Props, Subscription, Value>(
+  config: {|
+    // Specifies the name of the subscribable property.
+    // The subscription value will be passed along using this same name.
+    // In the case of a functional component, it will be passed as a `prop` with this name.
+    // For a class component, it will be set in `state` with this name.
+    property: string,
+
+    // Synchronously get the value for the subscribed property.
+    // Return undefined if the subscribable value is undefined,
+    // Or does not support synchronous reading (e.g. native Promise).
+    getValue: (props: Props) => Value,
+
+    // Setup a subscription for the subscribable value in props.
+    // Due to the variety of change event types, subscribers should provide their own handlers.
+    // Those handlers should not attempt to update state though;
+    // They should call the valueChangedCallback() instead when a subscription changes.
+    // You may optionally return a subscription value to later unsubscribe (e.g. event handler).
+    subscribe: (
+      props: Props,
+      valueChangedCallback: (value: Value) => void,
+    ) => Subscription,
+
+    // Unsubsribe from the subscribable value in props.
+    // The subscription value returned from subscribe() is passed as the second parameter.
+    unsubscribe: (props: Props, subscription: Subscription) => void,
+  |},
+  Component: React$ComponentType<*>,
+): React$ComponentType<Props> {
+  invariant(Component != null, 'Invalid subscribable Component specified');
+  invariant(
+    typeof config.property === 'string' && config.property !== '',
+    'Subscribable config must specify a subscribable property',
+  );
+  invariant(
+    typeof config.getValue === 'function',
+    'Subscribable config must specify a getValue function',
+  );
+  invariant(
+    typeof config.subscribe === 'function',
+    'Subscribable config must specify a subscribe function',
+  );
+  invariant(
+    typeof config.unsubscribe === 'function',
+    'Subscribable config must specify a unsubscribe function',
+  );
```

Comment:
Let's make these warnings instead. So bloated to have this at runtime/startup time.

Reply:
The component will runtime error without any of these properties, so I don't understand why a warning would be more appropriate.

Reply:
A warning wouldn't need this check in production code path. It would error in either case but for normal cases (where it doesn't error) it wouldn't need to spend CPU cycles on this.

Reply:
That makes sense. 👍